File: mind_snag_py/src/mind_snag/sorting/extract_spikes.py
"""Extract spike times from Kilosort output with drift correction.

Ports extract_spikes.m. Loads KS results, applies two-stage drift correction
(AP -> NIDQ -> recording timebase), and saves structured cluster data.
"""

from __future__ import annotations

import logging
from pathlib import Path

import numpy as np

from mind_snag.config import MindSnagConfig
from mind_snag.io.ks_loader import load_ks_dir
from mind_snag.io.cluster_groups import read_cluster_groups
from mind_snag.io.mat_reader import load_mat
from mind_snag.io.hdf5_writer import write_npclu_h5
from mind_snag.utils.channel_info import clus_channel_info
from mind_snag.utils.paths import (
    ks_output_dir, rec_name_str, group_flag_str, group_rec_dir,
)

logger = logging.getLogger(__name__)


def extract_spikes(
    cfg: MindSnagConfig,
    day: str,
    recs: list[str],
    tower: str,
    np_num: int,
    grouped: bool,
) -> None:
    """Extract spike times with drift correction from Kilosort output.

    Parameters
    ----------
    cfg : pipeline configuration
    day : recording date (YYMMDD)
    recs : list of recording numbers
    tower : recording setup name
    np_num : probe number
    grouped : whether recordings are concatenated
    """
    cfg.validate()
    data_root = cfg.data_root

    rec_str = rec_name_str(recs, grouped)
    gflag = group_flag_str(grouped)
    ks_dir = ks_output_dir(data_root, day, tower, np_num, rec_str, cfg.ks_version)

    if not ks_dir.exists():
        raise FileNotFoundError(
            f"Kilosort output directory not found: {ks_dir}\nRun run_kilosort4 first."
        )

    # Load KS output
    sp = load_ks_dir(ks_dir, exclude_noise=False)
    max_site, _ = clus_channel_info(cfg, day, recs, tower, np_num, grouped)

    # Load group file
    grp_dir = group_rec_dir(data_root, day, tower, np_num)
    group_file = grp_dir / f"spike_sorting_rec_groups_{rec_str}.mat"

    if grouped:
        _extract_grouped(cfg, sp, max_site, day, recs, tower, np_num, gflag, ks_dir, data_root)
    else:
        _extract_single(cfg, sp, max_site, day, recs[0], tower, np_num, gflag, ks_dir, data_root, group_file)


def _apply_drift_correction(
    spike_times: np.ndarray,
    w_nidq: np.ndarray,
    w_rec: np.ndarray,
) -> np.ndarray:
    """Apply two-stage drift correction: AP -> NIDQ -> recording timebase.

    Parameters
    ----------
    spike_times : raw spike times
    w_nidq : [intercept, slope] for AP -> NIDQ correction
    w_rec : [intercept, slope] for NIDQ -> recording correction

    Returns
    -------
    Drift-corrected spike times
    """
    st_nidq = w_nidq[0] + w_nidq[1] * spike_times
    return w_rec[0] + w_rec[1] * st_nidq


def _extract_grouped(
    cfg, sp, max_site, day, recs, tower, np_num, gflag, ks_dir, data_root,
):
    """Extract spikes for grouped (concatenated) recordings."""
    theo_offset = 0.0

    for i_r, rec in enumerate(recs):
        rec_dir = data_root / day / rec
        output_dir = data_root / day / rec
        output_dir.mkdir(parents=True, exist_ok=True)

        # Load metadata
        ap_meta_file = rec_dir / f"rec{rec}.{tower}.{np_num}.ap_meta.mat"
        nidq_meta_file = rec_dir / f"rec{rec}.nidq_meta.mat"
        if not ap_meta_file.exists():
            raise FileNotFoundError(f"AP meta file not found: {ap_meta_file}")

        ap_meta_data = load_mat(ap_meta_file)
        ap_meta = ap_meta_data.get("ap_meta", {})
        nidq_meta_data = load_mat(nidq_meta_file)
        nidq_meta = nidq_meta_data.get("nidq_meta", {})

        nsamp = _get_field(ap_meta, "nsamp")
        fs = _get_field(ap_meta, "Fs")
        theo_dur = nsamp / fs

        # Select spikes in this recording's time window
        mask = (sp.st <= theo_offset + theo_dur) & (sp.st > theo_offset)
        st_select = sp.st[mask] - theo_offset

        # Drift correction
        w_nidq = np.atleast_1d(_get_field(ap_meta, "nidq_drift_model_weights"))
        w_rec = np.atleast_1d(_get_field(nidq_meta, "rec_drift_model_weights"))

        new_spike_times = _apply_drift_correction(st_select, w_nidq, w_rec)
        new_spike_templates = sp.spike_templates[mask]

        theo_offset += theo_dur

        # Build NPclu (1-indexed cluster IDs for MATLAB compat)
        npclu = np.column_stack([new_spike_times, new_spike_templates + 1])
        templates = sp.temps
        pc_feat = sp.pc_feat[mask, :3, :] if sp.pc_feat is not None else None
        temp_scaling = sp.temp_scaling_amps[mask]

        clu_id = np.unique(sp.clu) + 1
        clu_info = np.column_stack([clu_id, max_site[: len(clu_id)]])

        # KS good units
        ks_file = ks_dir / "cluster_KSLabel.tsv"
        cids, cgs = read_cluster_groups(ks_file)
        ks_clu_id = cids[cgs == 2] + 1
        ks_mask = np.isin(clu_info[:, 0], ks_clu_id)
        ks_clu_info = clu_info[ks_mask]

        # Save
        out_file = output_dir / f"rec{rec}.{tower}.{np_num}.{gflag}.NPclu.h5"
        logger.info("Saving: %s", out_file)
        write_npclu_h5(
            out_file,
            spike_times=npclu[:, 0],
            cluster_ids=npclu[:, 1].astype(np.int64),
            templates=templates,
            clu_info=clu_info,
            ks_clu_info=ks_clu_info,
            pc_feat=pc_feat,
            temp_scaling_amps=temp_scaling,
            ks_version=cfg.ks_version,
        )


def _extract_single(
    cfg, sp, max_site, day, rec, tower, np_num, gflag, ks_dir, data_root, group_file,
):
    """Extract spikes for a single (non-grouped) recording."""
    rec_dir = data_root / day / rec

    # Try to load AP meta with NP number, fallback without
    ap_meta_file = rec_dir / f"rec{rec}.{tower}.{np_num}.ap_meta.mat"
    nidq_meta_file = rec_dir / f"rec{rec}.nidq_meta.mat"

    if not ap_meta_file.exists():
        ap_meta_file = rec_dir / f"rec{rec}.{tower}.ap_meta.mat"

    ap_meta_data = load_mat(ap_meta_file)
    ap_meta = ap_meta_data.get("ap_meta", {})
    nidq_meta_data = load_mat(nidq_meta_file)
    nidq_meta = nidq_meta_data.get("nidq_meta", {})

    nsamp = _get_field(ap_meta, "nsamp")
    fs = _get_field(ap_meta, "Fs")
    theo_dur = nsamp / fs

    mask = (sp.st < theo_dur) & (sp.st > 0)
    st_select = sp.st[mask]

    # Drift correction
    w_nidq = np.atleast_1d(_get_field(ap_meta, "nidq_drift_model_weights"))
    w_rec = np.atleast_1d(_get_field(nidq_meta, "rec_drift_model_weights"))

    new_spike_times = _apply_drift_correction(st_select, w_nidq, w_rec)
    new_spike_templates = sp.spike_templates[mask]

    npclu = np.column_stack([new_spike_times, new_spike_templates + 1])
    templates = sp.temps
    pc_feat = sp.pc_feat[mask, :3, :] if sp.pc_feat is not None else None
    temp_scaling = sp.temp_scaling_amps[mask]

    clu_id = np.unique(sp.clu) + 1
    clu_info = np.column_stack([clu_id, max_site[: len(clu_id)]])

    # KS good units
    ks_file = ks_dir / "cluster_KSLabel.tsv"
    cids, cgs = read_cluster_groups(ks_file)
    ks_clu_id = cids[cgs == 2] + 1
    ks_mask = np.isin(clu_info[:, 0], ks_clu_id)
    ks_clu_info = clu_info[ks_mask]

    output_dir = data_root / day / rec
    output_dir.mkdir(parents=True, exist_ok=True)

    out_file = output_dir / f"rec{rec}.{tower}.{np_num}.{gflag}.NPclu.h5"
    logger.info("Saving: %s", out_file)
    write_npclu_h5(
        out_file,
        spike_times=npclu[:, 0],
        cluster_ids=npclu[:, 1].astype(np.int64),
        templates=templates,
        clu_info=clu_info,
        ks_clu_info=ks_clu_info,
        pc_feat=pc_feat,
        temp_scaling_amps=temp_scaling,
        ks_version=cfg.ks_version,
    )


def _get_field(obj, name, default=0):
    """Get a field from a dict or scipy mat_struct."""
    if isinstance(obj, dict):
        return obj.get(name, default)
    return getattr(obj, name, default)

File: sorting/extract_spikes.m
function extract_spikes(cfg, day, rec, tower, np, grouped)
% EXTRACT_SPIKES Extract spike times from Kilosort output with drift correction
%
%   extract_spikes(cfg, day, rec, tower, np, grouped)
%
%   Loads Kilosort results, applies two-stage drift correction
%   (AP -> NIDQ -> recording timebase), and saves structured cluster data.
%
%   Parameters:
%     cfg     - Configuration struct from mind_snag_config
%     day     - Recording date as string, YYMMDD format (e.g. '250224')
%     rec     - Recording number(s), string or cell array (e.g. '007' or {'007','009'})
%     tower   - Recording setup name (e.g. 'LPPC_LPFC_modularV1')
%     np      - Neuropixel probe number (1 or 2)
%     grouped - 0 for individual recordings, 1 for concatenated
%
%   Outputs:
%     Saves rec{rec}.{tower}.{np}.{GroupFlag}.NPclu.mat containing:
%       NPclu      - [spike_time, cluster_id] matrix (drift-corrected)
%       NPtemplate - Template shapes [nTemplates x nTimePoints x nChannels]
%       Clu_info   - [cluster_id, channel_index] mapping
%       KSclu_info - Good units only (KS label = 'good')
%       pcFeat     - PC features for each spike
%       tempScalingAmps - Template scaling amplitudes
%       KSversion  - Kilosort version number
%
%   See also: run_kilosort4, compute_isolation, mind_snag_config

data_root = cfg.data_root;
assert(~isempty(data_root), 'mind_snag:config', 'cfg.data_root must be set.');

if ~exist('grouped', 'var'); grouped = 0; end

if cfg.ks_version == 4
    ks_name = '_KS4';
else
    ks_name = '';
end

KSversionflag = (cfg.ks_version == 4);

group_rec_dir = fullfile(data_root, day, 'spikeglx_data', ...
    ['grouped_recordings.' tower '.' num2str(np)]);

if grouped
    rec_name = strjoin(rec, '_');
    GROUPEDFLAG = 'Grouped';
else
    rec_name = rec;
    GROUPEDFLAG = 'NotGrouped';
end

ks_dir = fullfile(group_rec_dir, ['group' rec_name ks_name]);
assert(exist(ks_dir, 'dir') == 7, 'mind_snag:io', ...
    'Kilosort output directory not found: %s\nRun run_kilosort4 first.', ks_dir);

params.excludeNoise = false;
sp = loadKSdir(ks_dir, params);

groupfile = fullfile(group_rec_dir, ['spike_sorting_rec_groups_' rec_name '.mat']);
[max_site, ~] = clus_channel_infor(cfg, day, rec, tower, np, grouped);

if exist(groupfile, 'file') && grouped == 1
    load(groupfile, 'grouped_recs');
    ngroups = length(grouped_recs);

    for igroup = 1:ngroups
        group = grouped_recs{igroup};
        theo_offset = 0;

        for iR = 1:length(group)
            rec_this = rec{iR};
            rec_dir = fullfile(data_root, day, rec_this);
            output_dir = fullfile(data_root, day, rec_this);

            if ~exist(output_dir, 'dir')
                mkdir(output_dir);
            end

            ap_meta_file = fullfile(rec_dir, ['rec' rec_this '.' tower '.' num2str(np) '.ap_meta.mat']);
            nidq_meta_file = fullfile(rec_dir, ['rec' rec_this '.nidq_meta.mat']);
            assert(exist(ap_meta_file, 'file') == 2, 'mind_snag:io', ...
                'AP meta file not found: %s', ap_meta_file);
            load(ap_meta_file, 'ap_meta');
            load(nidq_meta_file);

            clear l_st_select st_select
            theo_dur = ap_meta.nsamp ./ ap_meta.Fs;
            l_st_select = sp.st <= theo_offset + theo_dur & sp.st > theo_offset;
            st_select = sp.st(l_st_select) - theo_offset;
            w_nidq = ap_meta.nidq_drift_model_weights;

            if ~isfield(nidq_meta, 'rec_drift_model_weights')
                warning('mind_snag:drift', ...
                    'nidq_meta missing rec_drift_model_weights for rec %s. Skipping drift correction.', rec_this);
            else
                theo_offset = theo_offset + theo_dur;
            end

            w_rec = nidq_meta.rec_drift_model_weights;
            clear st_select_nidq new_spiketimes new_spikeTemplates
            st_select_nidq = w_nidq(1) + w_nidq(2) * st_select;
            new_spiketimes = w_rec(1) + w_rec(2) * st_select_nidq;
            new_spikeTemplates = sp.spikeTemplates(l_st_select);
            pcFeat = sp.pcFeat(l_st_select, 1:3, :);
            tempScalingAmps = sp.tempScalingAmps(l_st_select);

            clear NPclu
            NPclu = zeros(length(new_spiketimes), 2);
            NPclu(:,1) = new_spiketimes;
            NPclu(:,2) = new_spikeTemplates + 1;
            NPtemplate = sp.temps;
            clu_id = unique(sp.clu) + 1;
            channelNum_ind = max_site;

            ksFile = fullfile(ks_dir, 'cluster_KSLabel.tsv');
            [cids, cgs] = readClusterGroupsCSV(ksFile);
            KS_cluid = cids(cgs == 2) + 1;

            clear Clu_info KSclu_info
            Clu_info(:,1) = clu_id;
            Clu_info(:,2) = channelNum_ind(clu_id);
            KSclu_info(:,1) = clu_id;
            KSclu_info(:,2) = channelNum_ind;
            KSversion = cfg.ks_version;

            out_file = fullfile(output_dir, ['rec' rec_this '.' tower '.' num2str(np) '.' GROUPEDFLAG '.NPclu.mat']);
            fprintf('Saving: %s\n', out_file);
            save(out_file, 'NPclu', 'NPtemplate', 'Clu_info', 'KSclu_info', ...
                'pcFeat', 'tempScalingAmps', 'KSversion', '-v7.3');
        end
    end

elseif exist(groupfile, 'file') && grouped == 0
    load(groupfile, 'grouped_recs');

    rec_this = rec;
    rec_dir = fullfile(data_root, day, rec_this);

    try
        ap_meta_file = fullfile(rec_dir, ['rec' rec '.' tower '.' num2str(np) '.ap_meta.mat']);
        nidq_meta_file = fullfile(rec_dir, ['rec' rec '.nidq_meta.mat']);
        load(ap_meta_file, 'ap_meta');
        load(nidq_meta_file);
    catch
        ap_meta_file = fullfile(rec_dir, ['rec' rec '.' tower '.ap_meta.mat']);
        nidq_meta_file = fullfile(rec_dir, ['rec' rec '.nidq_meta.mat']);
        load(ap_meta_file, 'ap_meta');
        load(nidq_meta_file);
    end

    theo_offset = 0;
    clear l_st_select st_select
    theo_dur = ap_meta.nsamp ./ ap_meta.Fs;

    l_st_select = sp.st < theo_offset + theo_dur & sp.st > theo_offset;
    st_select = sp.st(l_st_select) - theo_offset;

    w_nidq = ap_meta.nidq_drift_model_weights;

    if ~isfield(nidq_meta, 'rec_drift_model_weights')
        warning('mind_snag:drift', ...
            'nidq_meta missing rec_drift_model_weights. Check procAlignSourcesToRos.');
    end

    w_rec = nidq_meta.rec_drift_model_weights;

    clear st_select_nidq new_spiketimes new_spikeTemplates
    st_select_nidq = w_nidq(1) + w_nidq(2) * st_select;
    new_spiketimes = w_rec(1) + w_rec(2) * st_select_nidq;
    new_spikeTemplates = sp.spikeTemplates(l_st_select);

    NPclu = zeros(length(new_spiketimes), 2);
    NPclu(:,1) = new_spiketimes;
    NPclu(:,2) = new_spikeTemplates + 1;

    NPtemplate = sp.temps;
    pcFeat = sp.pcFeat(l_st_select, 1:3, :);
    tempScalingAmps = sp.tempScalingAmps(l_st_select);

    clu_id = unique(sp.clu) + 1;
    channelNum_ind = max_site;
    Clu_info(:,1) = clu_id;
    Clu_info(:,2) = channelNum_ind;

    ksFile = fullfile(ks_dir, 'cluster_KSLabel.tsv');
    [cids, cgs] = readClusterGroupsCSV(ksFile);
    KS_cluid = cids(cgs == 2) + 1;
    mask = ismember(Clu_info(:,1), KS_cluid);
    KSclu_info(:,1) = clu_id(mask);
    KSclu_info(:,2) = channelNum_ind(mask);

    KSversion = cfg.ks_version;

    output_dir = fullfile(data_root, day, rec_this);
    if ~exist(output_dir, 'dir')
        mkdir(output_dir);
    end

    out_file = fullfile(output_dir, ['rec' rec_this '.' tower '.' num2str(np) '.' GROUPEDFLAG '.NPclu.mat']);
    fprintf('Saving: %s\n', out_file);
    save(out_file, 'NPclu', 'NPtemplate', 'Clu_info', 'KSclu_info', ...
        'pcFeat', 'tempScalingAmps', 'KSversion', '-v7.3');
end

end
